File: warehouse-navigator/core/src/snapshot.ts
// Кодек офлайн-снимка склада: ПК-клиент экспортирует склад в файл (.whnav.json),
// телефон импортирует и навигирует без сети. Чистый JSON + строгая валидация на
// входе (чужой/битый файл → понятная ошибка, а не падение навигатора).

import type { WarehouseSnapshot } from './store.js';

export const SNAPSHOT_VERSION = 1 as const;

export class SnapshotError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SnapshotError';
  }
}

// Сериализация снимка в строку файла. pretty=false — компактно (для QR/передачи).
export function serializeSnapshot(s: WarehouseSnapshot, pretty = true): string {
  return JSON.stringify(s, null, pretty ? 2 : 0);
}

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}
function num(v: unknown): v is number {
  return typeof v === 'number' && Number.isFinite(v);
}
function str(v: unknown): v is string {
  return typeof v === 'string';
}
function numOrNull(v: unknown): v is number | null {
  return v === null || num(v);
}

// Разбор и ВАЛИДАЦИЯ строки файла в снимок. Бросает SnapshotError при любой
// несостыковке схемы — навигатор должен отклонить чужой файл, а не работать наугад.
export function parseSnapshot(text: string): WarehouseSnapshot {
  let raw: unknown;
  try {
    raw = JSON.parse(text);
  } catch {
    throw new SnapshotError('Файл не является корректным JSON');
  }
  if (!isObject(raw)) throw new SnapshotError('Ожидался объект снимка');
  if (raw.version !== SNAPSHOT_VERSION) {
    throw new SnapshotError(`Неподдерживаемая версия снимка: ${String(raw.version)}`);
  }
  if (!isObject(raw.warehouse) || !str(raw.warehouse.id) || !str(raw.warehouse.name)) {
    throw new SnapshotError('Некорректный склад в снимке');
  }
  const arr = (v: unknown, field: string): unknown[] => {
    if (!Array.isArray(v)) throw new SnapshotError(`Поле «${field}» должно быть массивом`);
    return v;
  };
  const wid = raw.warehouse.id;

  const cells = arr(raw.cells, 'cells').map((c, i) => {
    if (!isObject(c) || !str(c.id) || !str(c.code) || !str(c.warehouseId))
      throw new SnapshotError(`Некорректная ячейка #${i}`);
    if (c.warehouseId !== wid)
      throw new SnapshotError(`Ячейка #${i} принадлежит другому складу`); // иначе молча терялась бы
    if (!numOrNull(c.posXM) || !numOrNull(c.posYM))
      throw new SnapshotError(`Некорректные координаты ячейки #${i}`);
    return {
      id: c.id,
      code: c.code,
      warehouseId: c.warehouseId,
      posXM: c.posXM as number | null,
      posYM: c.posYM as number | null,
    };
  });

  const products = arr(raw.products, 'products').map((p, i) => {
    if (!isObject(p) || !str(p.id) || !str(p.sku) || !str(p.name))
      throw new SnapshotError(`Некорректный товар #${i}`);
    if (!(p.barcode === null || str(p.barcode)))
      throw new SnapshotError(`Некорректный штрихкод товара #${i}`);
    return { id: p.id, sku: p.sku, name: p.name, barcode: (p.barcode ?? null) as string | null };
  });

  const placements = arr(raw.placements, 'placements').map((pl, i) => {
    if (!isObject(pl) || !str(pl.productId) || !str(pl.cellId) || !num(pl.quantity))
      throw new SnapshotError(`Некорректное размещение #${i}`);
    if (pl.quantity < 0) throw new SnapshotError(`Отрицательное количество в размещении #${i}`);
    return { productId: pl.productId, cellId: pl.cellId, quantity: pl.quantity };
  });

  const layout = arr(raw.layout, 'layout').map((r, i) => {
    if (!isObject(r) || !num(r.xM) || !num(r.yM) || !num(r.lengthM) || !num(r.widthM))
      throw new SnapshotError(`Некорректный элемент плана #${i}`);
    return {
      xM: r.xM,
      yM: r.yM,
      lengthM: r.lengthM,
      widthM: r.widthM,
      rotationDeg: num(r.rotationDeg) ? r.rotationDeg : 0,
      kind: str(r.kind) ? (r.kind as WarehouseSnapshot['layout'][number]['kind']) : 'rack',
    };
  });

  const anchors = arr(raw.anchors, 'anchors').map((a, i) => {
    if (!isObject(a) || !str(a.warehouseId) || !num(a.xM) || !num(a.yM) || !num(a.headingDeg))
      throw new SnapshotError(`Некорректный якорь #${i}`);
    return { warehouseId: a.warehouseId, xM: a.xM, yM: a.yM, headingDeg: a.headingDeg };
  });

  return {
    version: SNAPSHOT_VERSION,
    warehouse: { id: wid, name: raw.warehouse.name },
    cells,
    products,
    placements,
    layout,
    anchors,
  };
}

File: warehouse-navigator/qt-client/src/Snapshot.cpp
#include "Snapshot.hpp"

#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonArray>
#include <QHash>
#include <stdexcept>

namespace whnav {

namespace {

QJsonValue numOrNull(const std::optional<double>& v) {
    return v ? QJsonValue(*v) : QJsonValue(QJsonValue::Null);
}

std::optional<double> readNumOrNull(const QJsonValue& v) {
    if (v.isNull() || v.isUndefined()) return std::nullopt;
    if (!v.isDouble()) throw std::runtime_error("Ожидалось число или null");
    return v.toDouble();
}

double reqNum(const QJsonObject& o, const QString& k) {
    const QJsonValue v = o.value(k);
    if (!v.isDouble()) throw std::runtime_error(QString("Поле %1 должно быть числом").arg(k).toStdString());
    return v.toDouble();
}

QString reqStr(const QJsonObject& o, const QString& k) {
    const QJsonValue v = o.value(k);
    if (!v.isString()) throw std::runtime_error(QString("Поле %1 должно быть строкой").arg(k).toStdString());
    return v.toString();
}

} // namespace

QByteArray Snapshot::toJson(bool pretty) const {
    QJsonObject root;
    root["version"] = 1;
    root["warehouse"] = QJsonObject{{"id", warehouse.id}, {"name", warehouse.name}};

    QJsonArray jc;
    for (const auto& c : cells) {
        jc.append(QJsonObject{
            {"id", c.id}, {"code", c.code}, {"warehouseId", c.warehouseId},
            {"posXM", numOrNull(c.posXM)}, {"posYM", numOrNull(c.posYM)}});
    }
    root["cells"] = jc;

    QJsonArray jp;
    for (const auto& p : products) {
        jp.append(QJsonObject{
            {"id", p.id}, {"sku", p.sku}, {"name", p.name},
            {"barcode", p.barcode.isEmpty() ? QJsonValue(QJsonValue::Null) : QJsonValue(p.barcode)}});
    }
    root["products"] = jp;

    QJsonArray jpl;
    for (const auto& pl : placements) {
        jpl.append(QJsonObject{
            {"productId", pl.productId}, {"cellId", pl.cellId}, {"quantity", pl.quantity}});
    }
    root["placements"] = jpl;

    QJsonArray jl;
    for (const auto& r : layout) {
        jl.append(QJsonObject{
            {"xM", r.xM}, {"yM", r.yM}, {"lengthM", r.lengthM}, {"widthM", r.widthM},
            {"rotationDeg", r.rotationDeg}, {"kind", r.kind}});
    }
    root["layout"] = jl;

    QJsonArray ja;
    for (const auto& a : anchors) {
        ja.append(QJsonObject{
            {"warehouseId", a.warehouseId}, {"xM", a.xM}, {"yM", a.yM}, {"headingDeg", a.headingDeg}});
    }
    root["anchors"] = ja;

    return QJsonDocument(root).toJson(pretty ? QJsonDocument::Indented : QJsonDocument::Compact);
}

Snapshot Snapshot::fromJson(const QByteArray& bytes) {
    QJsonParseError err{};
    const QJsonDocument doc = QJsonDocument::fromJson(bytes, &err);
    if (err.error != QJsonParseError::NoError || !doc.isObject())
        throw std::runtime_error("Файл не является корректным JSON-снимком");
    const QJsonObject root = doc.object();
    if (root.value("version").toInt(-1) != 1)
        throw std::runtime_error("Неподдерживаемая версия снимка");

    Snapshot s;
    const QJsonObject wh = root.value("warehouse").toObject();
    s.warehouse.id = reqStr(wh, "id");
    s.warehouse.name = reqStr(wh, "name");

    for (const QJsonValue v : root.value("cells").toArray()) {
        const QJsonObject o = v.toObject();
        Cell c;
        c.id = reqStr(o, "id");
        c.code = reqStr(o, "code");
        c.warehouseId = reqStr(o, "warehouseId");
        c.posXM = readNumOrNull(o.value("posXM"));
        c.posYM = readNumOrNull(o.value("posYM"));
        s.cells.append(c);
    }
    for (const QJsonValue v : root.value("products").toArray()) {
        const QJsonObject o = v.toObject();
        Product p;
        p.id = reqStr(o, "id");
        p.sku = reqStr(o, "sku");
        p.name = reqStr(o, "name");
        p.barcode = o.value("barcode").isString() ? o.value("barcode").toString() : QString();
        s.products.append(p);
    }
    for (const QJsonValue v : root.value("placements").toArray()) {
        const QJsonObject o = v.toObject();
        Placement pl;
        pl.productId = reqStr(o, "productId");
        pl.cellId = reqStr(o, "cellId");
        pl.quantity = reqNum(o, "quantity");
        s.placements.append(pl);
    }
    for (const QJsonValue v : root.value("layout").toArray()) {
        const QJsonObject o = v.toObject();
        LayoutRect r;
        r.xM = reqNum(o, "xM");
        r.yM = reqNum(o, "yM");
        r.lengthM = reqNum(o, "lengthM");
        r.widthM = reqNum(o, "widthM");
        r.rotationDeg = o.value("rotationDeg").toDouble(0);
        r.kind = o.value("kind").isString() ? o.value("kind").toString() : QString("rack");
        s.layout.append(r);
    }
    for (const QJsonValue v : root.value("anchors").toArray()) {
        const QJsonObject o = v.toObject();
        Anchor a;
        a.warehouseId = reqStr(o, "warehouseId");
        a.xM = reqNum(o, "xM");
        a.yM = reqNum(o, "yM");
        a.headingDeg = reqNum(o, "headingDeg");
        s.anchors.append(a);
    }
    return s;
}

QVector<Product> Snapshot::searchProducts(const QString& q) const {
    const QString s = q.trimmed().toLower();
    if (s.isEmpty()) return products;
    QVector<Product> out;
    for (const auto& p : products) {
        if (p.name.toLower().contains(s) || p.sku.toLower().contains(s) ||
            p.barcode.toLower().contains(s))
            out.append(p);
    }
    return out;
}

QVector<ProductLocation> Snapshot::productLocation(const QString& productId) const {
    QHash<QString, const Cell*> cellById;
    for (const auto& c : cells) cellById.insert(c.id, &c);

    QVector<ProductLocation> out;
    QHash<QString, int> idxByCell; // cellId → индекс в out (агрегация)
    for (const auto& pl : placements) {
        if (pl.productId != productId || pl.quantity <= 0) continue;
        const auto it = cellById.constFind(pl.cellId);
        if (it == cellById.constEnd()) continue;
        const Cell* c = it.value();
        const auto idxIt = idxByCell.constFind(c->id);
        if (idxIt != idxByCell.constEnd()) {
            out[idxIt.value()].quantity += pl.quantity;
        } else {
            idxByCell.insert(c->id, out.size());
            out.append(ProductLocation{c->id, c->code, c->posXM, c->posYM, pl.quantity});
        }
    }
    return out;
}

QString Snapshot::anchorPayload(const Anchor& a) {
    // Формат строго совпадает с core/parseAnchor: SSNAV1|warehouseId|xM|yM|headingDeg.
    return QString("SSNAV1|%1|%2|%3|%4")
        .arg(a.warehouseId)
        .arg(a.xM)
        .arg(a.yM)
        .arg(a.headingDeg);
}

} // namespace whnav
